stdlibs/stdlibs.go

```diff
@@ -275,6 +275,48 @@ func InjectPackage(store gno.Store, pn *gno.PackageNode) {
 				m.PushValue(res0)
 			},
 		)
+		pn.DefineNative("ExecAsPkg",
+			// TODO: rename UnsafeExecAsPkg?
```

Comment:
I'm still uncertain, so I'd appreciate additional input. However, it appears that we have three primary options.
* `std.ExecAsPkg`
* `std.UnsafeExecAsPkg`
* `unsafe.ExecAsPkg`

I believe it is essential to reassess our understanding of safety and risk when drafting contracts. After all, isn't the default assumption that everything is potentially unsafe?

Reply:
I havn't removed your comments.

But i think we should keep `std.ExecAsPkg`, it's a "normal feature" for a smart-contract engine to be able to do actions as the smart-contract addr

The use of the word `Unsafe` could scare contract writers.

Reply:
> I havn't removed your comments.
> 
> But i think we should keep `std.ExecAsPkg`, it's a "normal feature" for a smart-contract engine to be able to do actions as the smart-contract addr
> 
> The use of the word `Unsafe` could scare contract writers, and make them think as a bad practice.

@albttx 
Gnolang support pass function as parameters and interface callback. It does exactly what current std.ExecAsPkg's implementation tries to do, and is more flexible and safer.

Current std.ExecAsPkg implementation can only allow passing two parameters in a  function and no type checking on the function parameters. 

pn.DefineNative("ExecAsPkg",
			// TODO: rename UnsafeExecAsPkg?
			gno.Flds( // params
				"fn", gno.FuncT(nil, nil),
			),

https://github.com/gnolang/gno/blob/d59c320fce255e0add72eb2105608cfaac92d59e/stdlibs/stdlibs.go#L281

That is limited and unsafe. 

On the other hand, Gnolang, as in go, supports the passing functions as parameters with any number of parameters and does type checking of the passed functions' parameter types

To solve #634, as you mentioned, you can modify GRC20 with interface callback or pass a callback function as a parameter to allow the airdrop contract to alter the state of your token contract. In addition, you can add a static allowed list or even try to verify interface types for approval.

That is more idiomatic in go programming, safe and auditable.

Reply:
https://github.com/gnolang/gno/blob/d59c320fce255e0add72eb2105608cfaac92d59e/tests/files/zrealm_std6.gno#L21

Conclusion:

- This is not a valid test. All three lines should output the same address
- Meanwhile, it exposes an issue that breaks the readability of the code.
- More importantly, It makes checking std.GetOrigCaller() is losing its purpose in every contracts that require to check original caller.

Here is why: 

std.GetOrigCaller() suppose to get the first caller's address of the entire call path. Usually, it is end users who initiate a function call on a smart contract. In Ethereum, that is an external owned account (EOA). In gno, we do not separate it since the main package can also be the OrigCaller.

No matter where std.GetOrigCaller() is called in the call path; it should always return the same address. 

However, std.ExecAsPkg changes the context of the call path in the middle. It makes std.GetOrigCaller() returns the address of the contract that std.ExecAsPkg is called in. 

https://github.com/gnolang/gno/blob/d59c320fce255e0add72eb2105608cfaac92d59e/stdlibs/stdlibs.go#L308

    // Push an alternate context with OrigCaller=OrigPkgAddr, this will
    // affect the function call above
	m.PushOp(gno.OpSetContext)
	ctx := m.Context.(ExecContext)
	ctx.OrigCaller = ctx.OrigPkgAddr

It makes checking std.GetOrigCaller() is losing its purpose in every contracts that require to check original caller.